API/CopilotService/__init__.py

```diff
@@ -151,16 +161,28 @@ async def _build_task_context(raw_task, token):
     return task, chunks
 
 
-def _add_context_to_queston(context, question, task_list, usage_context):
+def _add_context_to_queston(context, question, task_list, usage_context, token_limit):
     context.custom_context.task_list_lens = len(task_list)
     context.custom_context.estimated_task_list_tokens = num_tokens_from_message(task_list)
     context.custom_context.estimated_usage_context_tokens = num_tokens_from_message(usage_context)
     if task_list:
         guiding_steps_separation = "\nHere are the steps you can refer to for this question:\n"
-        question = question + guiding_steps_separation + str(task_list) + '\n'
+        question = _try_add_steps_to_queston(question, guiding_steps_separation, task_list, token_limit)
     if usage_context:
         commands_info_separation = "\nBelow are some potentially relevant CLI commands information as context, please select the commands information that may be used in the scenario of the question from context, and supplement the missing commands information of context\n"
-        question = question + commands_info_separation + str(usage_context)
+        question = _try_add_steps_to_queston(question, commands_info_separation, usage_context, token_limit)
```

Comment:
Have we discard the chunks with the lowest correlation score from knowledge index service when the token exceeds the limit?

Reply:
The service uses `learn knowledge` to sort the retrieved chunks by similarity scores. It keeps the sorted order, but filters out some irrelevant chunks. Also, we haven't found a way to resort the chunks efficiently and accurately.

Reply:
Yes, I am considering whether this is the time to introduce the fine sorting.